### PR TITLE
PaintObject in json file

### PR DESCRIPTION
**Pull request by frutiemax**
I have a working circus painting using a json file format. It's considerably faster than my previous experiments, even without optimization.

The json file format defines:

- Some track sequence mapping tables (the circus example defines one table for the FlatTrack3x3)
- Some edge tables used to draw fences (the circus example defines one table for the FlatTrack3x3)
- Some "height support tables", those are used for defining the height of the tile elements (the circus defines one table)
- Some paint structs, those are selected with the given keys (trackElement, trackSequence and direction, in the future some keys will be added i.e. the vehicle properties pitch, num_peeps, etc.). The paint struct is drawn if the keys are respected, if one key is omited, it acts like a wildcard and is ignored. 

The paint struct can draw:
- Supports
- Floor
- Fences
- AddImageAsParent
- SetSegmentSupportHeight

[twist_paint.zip](https://github.com/OpenRCT2/OpenRCT2/files/10407827/twist_paint.zip)
[circus_paint.zip](https://github.com/OpenRCT2/OpenRCT2/files/10407829/circus_paint.zip)

### CODE REVIEW

**Comment by IntelOrca** [comment]
The foundation looks good. Being able to define these individual paint structs provides good flexability, but we probably also want to allow the JSON file to specify a common base - for example the Circus and Twister would ask for the same base, a cork floor with fence and wooden supports - then use specific paint structs for the structure.

For flat rides, would it be simpler to just have this information in the ride type object? I can't see why another ride type would share the paint code for Circus (unlike tracks which are shared) - so it makes more sense to put this in the circus ride object. The vehicle objects (current ride objects) would then contain the ratings, sprites and strings.

**Comment by frutiemax** [comment]
> The foundation looks good. Being able to define these individual paint structs provides good flexability, but we probably also want to allow the JSON file to specify a common base - for example the Circus and Twister would ask for the same base, a cork floor with fence and wooden supports - then use specific paint structs for the structure.
> 
> For flat rides, would it be simpler to just have this information in the ride type object? I can't see why another ride type would share the paint code for Circus (unlike tracks which are shared) - so it makes more sense to put this in the circus ride object. The vehicle objects (current ride objects) would then contain the ratings, sprites and strings.

You'd still want to support flat rides with different shapes other than 3x3. With this, you can define the sequence mapping for 10x5 for example, as long as the TrackElement is defined in the RTD... Also, the flat ride floor painting is optional, you can just not put the paint struct to not paint it. You also don't need to define a paint struct for every key combination (direction, trackSequence, etc.) since it is always needed to be drawn for circus. I don't think it adds too much complexity...

**Comment by IntelOrca** [comment]
> You'd still want to support flat rides with different shapes other than 3x3. With this, you can define the sequence mapping for 10x5 for example, as long as the TrackElement is defined in the RTD... Also, the flat ride floor painting is optional, you can just not put the paint struct to not paint it. You also don't need to define a paint struct for every key combination (direction, trackSequence, etc.) since it is always needed to be drawn for circus. I don't think it adds too much complexity...

Of course, I was just thinking we could define built-in presets for objects to use.

**Comment by frutiemax** [comment]
The presets would be hardcoded or an another object type?

**Comment by IntelOrca** [comment]
@frutiemax hardcoded is much simpler to implement for now

**Comment by frutiemax** [comment]
I got rid of the TrackSequenceMapping table for the paint structs, it was a bit a puzzle to figure out the good numbers...
[object.txt](https://github.com/OpenRCT2/OpenRCT2/files/10369204/object.txt)

I will need to make some kind of debugging code to get the good key/values combinations as the animated rides is near impossible to do the json by hand.

**Comment by frutiemax** [comment]
[twist_paint.zip](https://github.com/OpenRCT2/OpenRCT2/files/10405927/twist_paint.zip)

The Twist paint code is working.

File: src/openrct2/paint/TreeContainer.h [hunk 1]
```diff
@@ -106,7 +106,8 @@ template<class KeyType, class ValueType> void TreeContainer<KeyType, ValueType>:
         const auto& key = keyValue.first;
         const auto& value = keyValue.second;
 
-        auto keysGenerated = _keyGenerator.GenerateKeys(key);
+        std::vector<KeyType> keysGenerated;
+        _keyGenerator.GenerateKeys(key, keysGenerated);
```
Comment (ZehMatt) [inline src/openrct2/paint/TreeContainer.h:110]: This makes no difference performance wise if thats the reason to why you changed it, returning complex objects such as vector is already handled internally as the first parameter, see https://godbolt.org/z/fvYTe57Yq

Comment (ZehMatt) [reply src/openrct2/paint/TreeContainer.h:110]: You can gain a bit of performance if you move the vector out of the for loop scope and call clear before calling GenerateKeys, that will avoid allocations as clear will only resize but keeps the capacity.

Comment (frutiemax) [reply src/openrct2/paint/TreeContainer.h:110]: I'm more concerned about the RAM usage currently... If you have any idea on how to reduce that problem.

Comment (frutiemax) [reply src/openrct2/paint/TreeContainer.h:110]: The RAM problem is at the loading, it gets down afterward.

Comment (frutiemax) [reply src/openrct2/paint/TreeContainer.h]: [motion_simulator.zip](https://github.com/OpenRCT2/OpenRCT2/files/10495451/motion_simulator.zip)
[twist_paint.zip](https://github.com/OpenRCT2/OpenRCT2/files/10495452/twist_paint.zip)
[enterprise.zip](https://github.com/OpenRCT2/OpenRCT2/files/10495453/enterprise.zip)
Try with these in the object folder.

Comment (ZehMatt) [reply src/openrct2/paint/TreeContainer.h]: Well as the godbolt demo illustrates changing the function signature like that will have no effect at all besides making it harder to read. I would return the vector if its not supposed to be re-used in any way.

**Comment by ZehMatt** [comment]
@frutiemax are you planning to continue on this?

**Pull request closed without merging**